### PR TITLE
Bug 1745192: Move registries.conf editing into a subpackage

### PR DESCRIPTION
**Pull request by mtrmac**
<s>(WIP because this includes unmerged https://github.com/openshift/machine-config-operator/pull/1082 .)</s>

**- What I did**
Turned `updateRegistriesConfig` from `pkg/controller/container-runtime-config` into 
`pkg/registries.EditRegistriesConfig` in a separate subpackage, with minimal dependencies
to be possible to call it from other registries.

This is in preparation for possibly moving it out of this repository into library-go entirely, see https://github.com/openshift/openshift-controller-manager/pull/19 for the whole discussion.

Moving it only into a subpackage in this repo is clearly non-disruptive, and it allows reviewing the proposed API before moving it.

The API deals with Go structures only, forcing the callers to do their own TOML encoding/decoding. This is to make it easier for the openshift-controller-manager code to start with a Go-only data structure without having to encode it just to submit it to the edit function.

**- How to verify it**
Existing unit tests.

**- Description for the changelog**
N/A

### CODE REVIEW

**Comment by mtrmac** [comment]
@runcom @umohnani8 @gabemontero PTAL.

**Comment by mtrmac** [comment]
@kikisdeliveryservice No, https://github.com/openshift/openshift-controller-manager/pull/19 needs to happen in 4.2 ; this is to avoid duplicating all of that code in two places.

We are still looking for the best place for the newly created subpackage to exist, MCO is not ideal.

File: pkg/registries/topo.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package containerruntimeconfig
+package registries
```
Comment (runcom) [inline pkg/registries/topo.go]: can this pkg be under `pkg/controller/container-runtime-config`?

Comment (mtrmac) [reply pkg/registries/topo.go]: Sure; I moved it up here to make it easier to find / read for out-of-repo consumers (who don’t care about the M-C-controller much).

Ideally we would want to have it somewhere out of the 2 (3? 4?) known users into some shared infrastructure repo.

If it is going to stay in MCO, I really don’t care where.

**Comment by mtrmac** [comment]
(Oh, and I forgot, just in case: See the individual commits to see what changed.)

File: pkg/registries/topo.go (hunk 1 above)
Comment (mtrmac) [reply pkg/registries/topo.go]: Moved.

**Comment by mtrmac** [comment]
Filed https://github.com/openshift/library-go/pull/513 for the library-go inclusion; that would replace this PR.

File: pkg/controller/container-runtime-config/helpers.go [hunk 2]
```diff
@@ -241,164 +240,21 @@ func updateCRIOConfig(data []byte, internal *mcfgv1.ContainerRuntimeConfiguratio
...
-		}
-		tomlConf.Registries = append(tomlConf.Registries, sysregistriesv2.Registry{
-			Endpoint: sysregistriesv2.Endpoint{Location: scope},
-		})
-		return &tomlConf.Registries[len(tomlConf.Registries)-1]
-	}
-
-	mirrorSets, err := mergedMirrorSets(icspRules)
-	if err != nil {
+	if err := registries.EditRegistriesConfig(&tomlConf, internalInsecure, internalBlocked, icspRules); err != nil {
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/controller/container-runtime-config/helpers.go]: I would be very surprised if that helps; a library-go subpackage with that modified API would pull in _exactly the same_ containers/image, and other, dependencies, as the code calling it does right now, there would just be one layer of indirection.

The effect would basically be only that library-go would _always_ pull in the TOML encoder/decoder as well (which makes no difference for the two currently contemplated consumers, but removes a bit of flexibility for possible future uses).

---

One thing that could possibly actually help is fixing https://github.com/containers/image/issues/692 ; my problems with Glide in the library-go PR manifested in one of the compression implementations (e.g. via `github.com/klauspost/compress/snappy/testdata/alice29.txt` ).

But that would help with either of the APIs, and might not be essential — I was able to make Glide work with that one package using `git add --force`, at least.

**Comment by gabemontero** [comment]
I've sent out an email to the interested parties on getting to a decision on merging this PR and then rebasing https://github.com/openshift/openshift-controller-manager/pull/19 off of this

**Comment by umohnani8** [comment]
LGTM
@runcom PTAL

**Comment by runcom** [comment]
> LGTM
> @runcom PTAL

we need a BZ explaining why we need such PR

**Comment by umohnani8** [comment]
The BZ for this is https://bugzilla.redhat.com/show_bug.cgi?id=1745192. Well it requires https://github.com/openshift/openshift-controller-manager/pull/19 as well to be fully fixed.

**Comment by runcom** [comment]
/approve

will leave to @kikisdeliveryservice one last review

**Pull request merged**